src/treelearner/serial_tree_learner.cpp

```diff
@@ -804,51 +774,43 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
   }
 }
 
-
-// this function updates the best split for a leaf
-// it is called only when monotone constraints exist
-void SerialTreeLearner::UpdateBestSplitsFromHistograms(
-    SplitInfo *split, int leaf,
-    const std::vector<int8_t> &is_feature_used_,
-    int num_features_, HistogramPool &histogram_pool_,
-    LearnerState &learner_state) {
-  // the feature histogram is retrieved
-  FeatureHistogram *histogram_array_;
-  histogram_pool_.Get(leaf, &histogram_array_);
+void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
+  FeatureHistogram* histogram_array_;
+  if (!histogram_pool_.Get(leaf, &histogram_array_)) {
+    Log::Warning(
+        "Get historical Histogram for leaf %d failed, will skip the "
+        "``RecomputeBestSplitForLeaf``",
+        leaf);
+    return;
+  }
   double sum_gradients = split->left_sum_gradient + split->right_sum_gradient;
   double sum_hessians = split->left_sum_hessian + split->right_sum_hessian;
   int num_data = split->left_count + split->right_count;
-  split->gain = kMinScore;
+
+  std::vector<SplitInfo> bests(num_threads_);
+  LeafSplits leaf_splits(num_data);
+  leaf_splits.Init(leaf, sum_gradients, sum_hessians);
 
   OMP_INIT_EX();
-#pragma omp parallel for schedule(static, 1024) if (num_features_ >= 2048)
+// find splits
+#pragma omp parallel for schedule(static)
   for (int feature_index = 0; feature_index < num_features_; ++feature_index) {
     OMP_LOOP_EX_BEGIN();
-    // splits are computed for the feature that are to be used
-    if (!histogram_array_[feature_index].is_splittable()) {
```

Comment:
Just to make sure do we not need to check is the histogram array is splittable here?

Reply:
We actually need to check for that otherwise I can get splits without any data on one side.